File: app/src/main/java/com/examle/curexchange/ui/adapter/CurrencyAdapter.java
package com.examle.curexchange.ui.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.examle.curexchange.R;

import java.util.List;

public class CurrencyAdapter extends RecyclerView.Adapter<CurrencyAdapter.ViewHolder> {

    private final List<String> currenciesList;
    private OnRecyclerItemClickedListener listener;

    public CurrencyAdapter(List<String> currencyList, OnRecyclerItemClickedListener listener) {
        this.currenciesList = currencyList;
        this.listener = listener;
    }

    public void setData(List<String> currenciesList) {
        this.currenciesList.clear();
        this.currenciesList.addAll(currenciesList);
        notifyDataSetChanged();
    }

    @Override
    public CurrencyAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.currency_item_view, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final CurrencyAdapter.ViewHolder holder, int position) {
        holder.currencyNameTextView.setText(currenciesList.get(position));
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onClicked(holder.currencyNameTextView.getText().toString());
            }
        });
    }

    @Override
    public int getItemCount() {
        return currenciesList.size();
    }

    public interface OnRecyclerItemClickedListener {

        void onClicked(String name);
    }

    protected static class ViewHolder extends RecyclerView.ViewHolder {

        private TextView currencyNameTextView;

        public ViewHolder(View itemView) {
            super(itemView);
            currencyNameTextView = itemView.findViewById(R.id.currency_name);
        }
    }
}

File: app/src/main/java/com/examle/curexchange/ui/adapter/CurrencyAdapter.kt
package com.examle.curexchange.ui.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.examle.curexchange.R
import kotlinx.android.synthetic.main.currency_item_view.view.*

class CurrencyAdapter(private val currencyList: MutableList<String>,
                      private val listener: OnRecyclerItemClickListener)
    : RecyclerView.Adapter<CurrencyAdapter.ViewHolder>() {

    fun setData(currencyList: MutableList<String>) {
        this.currencyList.clear()
        this.currencyList.addAll(currencyList)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater
                .from(parent.context)
                .inflate(R.layout.currency_item_view, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(currencyList[position], listener)
    }

    override fun getItemCount(): Int {
        return currencyList.size
    }

    interface OnRecyclerItemClickListener {
        fun onClicked(name: String)
    }


    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(currencyName: String, listener: OnRecyclerItemClickListener) = with(itemView) {
            itemView.currency_name.text = currencyName
            itemView.setOnClickListener { listener.onClicked(currencyName) }
        }
    }
}